src/validation.cpp

```diff
@@ -399,6 +399,35 @@ void UpdateMempoolForReorg(DisconnectedBlockTransactions &disconnectpool, bool f
     LimitMempoolSize(mempool, GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000, GetArg("-mempoolexpiry", DEFAULT_MEMPOOL_EXPIRY) * 60 * 60);
 }
 
+// Used to avoid mempool polluting consensus critical paths if CCoinsViewMempool
+// were somehow broken and returning the wrong scriptPubKeys
+static bool CheckInputsFromMempoolAndCache(const CTransaction& tx, CValidationState &state, const CCoinsViewCache &view, CTxMemPool& pool,
+                 unsigned int flags, bool cacheSigStore, PrecomputedTransactionData& txdata) {
+    AssertLockHeld(cs_main);
+
+    assert(!tx.IsCoinBase());
+    {
+        LOCK(pool.cs);
+        for (const CTxIn& txin : tx.vin) {
+            const Coin& coin = view.AccessCoin(txin.prevout);
+            if (coin.IsSpent()) return false;
```

Comment:
> I believe the code is correct. No entry was found in the cache, so validation should fail - it's not a consistency bug.

I also think the code is correct, just would like a comment saying why this is a return rather than a continue, or something that asserts against the CheckInputs() result. The whole function except this one line seems intended to do a bunch of checks and catch logic errors, while this line bypasses checks and seems to be doing some kind of optimization. AFAICT. Right now the intent is unclear.